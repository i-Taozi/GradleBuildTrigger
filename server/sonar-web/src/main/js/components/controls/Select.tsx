/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import styled from '@emotion/styled';
import * as React from 'react';
import ReactSelect, { GroupTypeBase, IndicatorProps, Props, StylesConfig } from 'react-select';
import { colors, others, sizes, zIndexes } from '../../app/theme';

const ArrowSpan = styled.span`
  border-color: #999 transparent transparent;
  border-style: solid;
  border-width: 4px 4px 2px;
  display: inline-block;
  height: 0;
  width: 0;
`;

export default function Select<
  Option,
  IsMulti extends boolean = false,
  Group extends GroupTypeBase<Option> = GroupTypeBase<Option>
>(props: Props<Option, IsMulti, Group>) {
  function DropdownIndicator({ innerProps }: IndicatorProps<Option, IsMulti, Group>) {
    return <ArrowSpan {...innerProps} />;
  }

  return (
    <ReactSelect
      {...props}
      styles={selectStyle<Option, IsMulti, Group>()}
      components={{
        ...props.components,
        DropdownIndicator
      }}
    />
  );
}

export function selectStyle<
  Option,
  IsMulti extends boolean,
  Group extends GroupTypeBase<Option>
>(): StylesConfig<Option, IsMulti, Group> {
  return {
    container: () => ({
      position: 'relative',
      display: 'inline-block',
      verticalAlign: 'middle',
      fontSize: '12px',
      textAlign: 'left'
    }),
    control: () => ({
      position: 'relative',
      display: 'table',
      width: '100%',
      height: `${sizes.controlHeight}`,
      lineHeight: `calc(${sizes.controlHeight} - 2px)`,
      border: `1px solid ${colors.gray80}`,
      borderCollapse: 'separate',
      borderRadius: '2px',
      backgroundColor: '#fff',
      boxSizing: 'border-box',
      color: `${colors.baseFontColor}`,
      cursor: 'default',
      outline: 'none'
    }),
    singleValue: () => ({
      bottom: 0,
      left: 0,
      lineHeight: '23px',
      paddingLeft: '8px',
      paddingRight: '24px',
      position: 'absolute',
      right: 0,
      top: 0,
      maxWidth: '100%',
      overflow: 'hidden',
      textOverflow: 'ellipsis',
      whiteSpace: 'nowrap'
    }),
    valueContainer: () => ({
      bottom: 0,
      left: 0,
      lineHeight: '23px',
      paddingLeft: '8px',
      paddingRight: '24px',
      position: 'absolute',
      right: 0,
      top: 0,
      maxWidth: '100%',
      overflow: 'hidden',
      textOverflow: 'ellipsis',
      whiteSpace: 'nowrap'
    }),
    indicatorsContainer: () => ({
      cursor: 'pointer',
      display: 'table-cell',
      position: 'relative',
      textAlign: 'center',
      verticalAlign: 'middle',
      width: '20px',
      paddingRight: '5px'
    }),
    menu: () => ({
      borderBottomRightRadius: '4px',
      borderBottomLeftRadius: '4px',
      backgroundColor: '#fff',
      border: '1px solid #ccc',
      borderTopColor: `${colors.barBorderColor}`,
      boxSizing: 'border-box',
      marginTop: '-1px',
      maxHeight: '200px',
      position: 'absolute',
      top: '100%',
      width: '100%',
      zIndex: `${zIndexes.dropdownMenuZIndex}`,
      webkitOverflowScrolling: 'touch',
      boxShadow: `${others.defaultShadow}`
    }),
    menuList: () => ({
      maxHeight: '198px',
      padding: '5px 0',
      overflowY: 'auto'
    }),
    option: (_provided, state) => ({
      display: 'block',
      lineHeight: '20px',
      padding: '0 8px',
      boxSizing: 'border-box',
      color: `${colors.baseFontColor}`,
      backgroundColor: state.isFocused ? `${colors.barBackgroundColor}` : 'white',
      fontSize: `${sizes.smallFontSize}`,
      cursor: 'pointer',
      whiteSpace: 'nowrap',
      overflow: 'hidden',
      textOverflow: 'ellipsis'
    })
  };
}
