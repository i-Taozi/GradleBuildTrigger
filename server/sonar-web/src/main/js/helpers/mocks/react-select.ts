/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { GroupTypeBase, OptionProps } from 'react-select';

export function mockReactSelectOptionProps<
  OptionType,
  IsMulti extends boolean,
  GroupType extends GroupTypeBase<OptionType> = GroupTypeBase<OptionType>
>(
  data: OptionType,
  overrides?: OptionProps<OptionType, IsMulti, GroupType>
): OptionProps<OptionType, IsMulti, GroupType> {
  return {
    ...overrides,
    data
  } as OptionProps<OptionType, IsMulti, GroupType>;
}
