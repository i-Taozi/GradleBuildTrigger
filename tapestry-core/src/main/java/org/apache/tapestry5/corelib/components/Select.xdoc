<document>
    <body>
        <section name="Simple Example">

            <p>
                A simple example, selecting one of three strings:
            </p>

            <subsection name="SelectColor.tml">
                <source><![CDATA[
<html xmlns:t="http://tapestry.apache.org/schema/tapestry_5_0_0.xsd">
    <body>
        <h1>Select Color</h1>

        <t:form>

            <t:label for="color"/>:
            <t:select t:id="color" model="literal:Red,Green,Blue"/>

            <br/>

            <input type="submit" value="Continue"/>
        </t:form>

    </body>
</html>]]></source>

                <p>
                    When the model parameter is a string, it is split apart at the commas.  When the model parameter is
                    a List of Strings, each element is considered a selection option.
                </p>

                <p>
                    When using this approach, you'll commonly put the list into the message catalog,
                    and reference it using a "message:" binding.
                </p>

                <p>
                    Here the label displayed to the user is the same as the value used to update the property, but that
                    doesn't have to be the case. By specifying a value and a label, we can control the server side value
                    without changing how the UI appears to the user:
                </p>

                <source>
                    <![CDATA[
            <t:select t:id="color" model="literal:FF0000=Red,00FF00=Green,0000FF=Blue"/>
]]>
                </source>


            </subsection>

<subsection name="SelectColor.java">

<source><![CDATA[
public class SelectColor
{
  @Validate("required")
  public String color;

}]]></source>

    <p>
        Placing the @Validate annotation on the field supplies the default validation for this field, here
        it indicates that color is required.  This prevents the
        Select component from including a blank option for the field (though this behavior too can be overridden).
        Without the @Validate, it would be possible for the user to select a blank value, and null would
        be assigned to the color property of SelectColor.  This might be appropriate for a search form, but
        not for an edit form.
    </p>

</subsection>

        </section>

        <section name="Enum Example">

            <p>Working with Enums is, amazingly, even easier (and more so than with the Radio component).</p>

            <p>
                Most of this example is the same as for the
                <a href="Radio.html">Radio</a>
                component, except that
                we're using a single Select component instead of multiple Radio components:
            </p>

            <p>
                <img src="select_ref2.png"/>
            </p>

            <subsection name="CardType.java">

                <source><![CDATA[
public enum CardType
{
    MASTER_CARD, VISA, AMERICAN_EXPRESS, DINERS_CLUB, DISCOVER
}
]]></source>

            </subsection>

            <subsection name="Payment.tml (partial)">

                <p>
                    In the Radio example, we used a Label and a Radio component for each enum value.
                    With Select we use a single Label and a single Select component:
                </p>

                <source><![CDATA[
    <t:label for="type"/>:
    <t:select t:id="type"/>]]></source>


                <p>
                    Here again, Tapestry is binding the value parameter to the type property, based on the component's
                    id.
                    In addition, because the property type of the property in question is an enum, a SelectModel
                    is automatically generated.
                </p>

            </subsection>

            <subsection name="Payment.java (partial)">

                <source><![CDATA[                                                                   
public class Payment
{
    . . .

    @Property
    @Persist
    @Validate("required")
    private CardType type;


    . . .
}]]></source>


            </subsection>

            <subsection name="Payment.properties">

                <p>
                    Once again, we need to slightly customize Tapestry's guess at the label for the enum value
                    DINERS_CLUB.
                    In the Radio example, we overrode the label for the dinersClub<em>component</em>. Here there is just
                    the Select component,
                    but we still have an option: override how the DINERS_CLUB enum value is displayed:
                </p>

                <source><![CDATA[DINERS_CLUB=Diner's Club]]></source>

                <p>Tapestry looks inside a component's message catalog for entries before "humanizing" the name. In the
                    <em>very rare</em>
                    event that there's a naming conflict, you may qualify the enum value with its class name:
                </p>


                <source><![CDATA[CardType.DINERS_CLUB=Diner's Club]]></source>

                <p>And, of course, all of this is case insensitive. The use of case here helps to identify how the
                    values
                    are to be used.
                </p>
            </subsection>

        </section>
        
        <section name="Chaining of Select components">
        	There is often a requirement for chaining Select components. When a value of a Select component is changed 
        	another Select should become visible. Let's consider the following example: you create an online shop for a car seller.
        	A make is modeled as enumeration CarMaker.  
        	
        	<subsection name="CarMaker.java">
                <source><![CDATA[                                                                   
public enum CarMaker
{
    MERCEDES, AUDI, BMW;
}]]></source>  
			</subsection>
			
			<subsection name="SelectZoneDemo.tml">
				The Select component 'carMaker' of the page SelectZoneDemo shows all available car makers. 
				When a user selects a car maker, another Select component for selecting available models of the make should appear.
				This can be accomplished with the <em>zone</em> parameter of the 'carMaker' Select component. 
				When the <em>zone</em> parameter is bound, every change of the Select's value causes an Ajax
				request, firing the <em>valueChanged</em> event. In this example, a corresponding event handler
				method, onValueChanged, populates the <em>model</em> used by the second Select component.
       	                <source><![CDATA[                                                                   
<html xmlns:t="http://tapestry.apache.org/schema/tapestry_5_1_0.xsd">
   <t:form>
      <p>
         <t:errors />
      </p>
      <p>
         <t:select t:id="carMaker" validate="required"
                   zone="modelZone" />
      </p>
      
      <t:zone t:id="modelZone" id="modelZone">
         <t:if test="carMaker">
           <p>
              <t:select t:id="carModel" model="availableModels" validate="required"/>
           </p>
         </t:if>
      </t:zone>
      
      <p>
         <t:submit value="literal:Submit" />
      </p>
   </t:form>

</html>]]></source>
        </subsection>
        
	    <subsection name="SelectZoneDemo.java">
        		The event handler method for the event <em>valuechanged</em> is used to provide the available car models of the currently selected car maker.
        		The new Select's value is passed as context. 
        	                <source><![CDATA[                                                                   
public class SelectZoneDemo
{
   
   @Inject
   private Messages messages;
   
   @Property
   @Persist
   private CarMaker carMaker;
   
   @Property
   @Persist
   private String carModel;

   @InjectComponent
   private Zone modelZone;
   
   @Property
   @Persist
   private List<String> availableModels;
    
    public Object onValueChanged(CarMaker maker) 
    {
       availableModels = findAvailableModels(maker);
       
       return modelZone.getBody();
    }
    
    public List<String> findAvailableModels(final CarMaker maker) 
    {
      switch (maker) 
      {
         case AUDI:
            return Arrays.asList("A4", "A6", "A8");
         case BMW:
            return Arrays.asList("3 Series", "5 Series", "7 Series");
         case MERCEDES:
            return Arrays.asList("C-Class", "E-Class", "S-Class");
         default:
            return Arrays.asList();
       }
    }    
}]]></source>


<p>
The Select component is very smart for enum types; it can automatically create a SelectModel for a given Enum, and a default
ValueEncoder.  Likewise, it can turn an array or List into a SelectModel automatically.  This streamlines the use of the Select
in many situations ... but because the model and encode parameters are still present, allows you to override its behavior
when needed.
</p>

			</subsection>
        
        </section>

    </body>
</document>