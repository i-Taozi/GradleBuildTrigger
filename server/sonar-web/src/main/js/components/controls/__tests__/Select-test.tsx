/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { shallow } from 'enzyme';
import * as React from 'react';
import { GroupTypeBase, Props } from 'react-select';
import Select from '../Select';

describe('Select', () => {
  it('should render correctly', () => {
    expect(shallowRender()).toMatchSnapshot('default');
  });

  function shallowRender<
    Option,
    IsMulti extends boolean = false,
    Group extends GroupTypeBase<Option> = GroupTypeBase<Option>
  >(props: Partial<Props<Option, IsMulti, Group>> = {}) {
    return shallow<Props<Option, IsMulti, Group>>(<Select {...props} />);
  }
});
