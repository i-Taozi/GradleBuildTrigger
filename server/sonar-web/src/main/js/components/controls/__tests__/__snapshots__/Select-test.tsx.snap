// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render correctly: default 1`] = `
<StateManager
  components={
    Object {
      "DropdownIndicator": [Function],
    }
  }
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  styles={
    Object {
      "container": [Function],
      "control": [Function],
      "indicatorsContainer": [Function],
      "menu": [Function],
      "menuList": [Function],
      "option": [Function],
      "singleValue": [Function],
      "valueContainer": [Function],
    }
  }
/>
`;
